sycl/include/CL/sycl/property_list.hpp

```diff
@@ -8,332 +8,123 @@
-
-class property_list {
-
-  // The structs validate that all objects passed are base of PropBase class.
-  template <typename... Tail> struct AllProperties : std::true_type {};
-  template <typename T, typename... Tail>
-  struct AllProperties<T, Tail...>
-      : std::conditional<std::is_base_of<property::detail::PropBase, T>::value,
-                         AllProperties<Tail...>, std::false_type>::type {};
-
-  template <class T>
-  using PropertyHolder = cl::sycl::property::detail::PropertyHolder<T>;
-  template <property::detail::PropKind PropKindT>
-  using Property = cl::sycl::property::detail::Prop<PropKindT>;
-
-  // The structs build tuple type that can hold all properties.
-  template <typename... Head> struct DefineTupleType {
-    using Type = std::tuple<Head...>;
-  };
-
-  template <int Counter, typename... Head>
-  struct BuildTupleType
-      : public std::conditional<
-            (Counter < property::detail::PropKind::PropKindSize),
-            BuildTupleType<
-                Counter + 1, Head...,
-                PropertyHolder<typename Property<(property::detail::PropKind)(
-                    Counter)>::FinalType>>,
-            DefineTupleType<Head...>>::type {};
-
-public:
-  // C'tor initialize m_PropList with properties passed by invoking ctorHelper
-  // recursively
-  template <typename... propertyTN,
-            typename = typename std::enable_if<
-                AllProperties<propertyTN...>::value>::type>
-  property_list(propertyTN... Props) {
+  template <typename... PropsT, class PropT>
+  typename std::enable_if<
+      std::is_base_of<detail::ComplexPropertyBase, PropT>::value>::type
+  ctorHelper(PropT &Prop, PropsT... Props) {
+    MComplexProps.emplace_back(new PropT(Prop));
     ctorHelper(Props...);
   }
 
-  template <typename propertyT> propertyT get_property() const {
-    if (!has_property<propertyT>()) {
-      throw sycl::invalid_object_error();
-    }
-    const auto &PropHolder =
-        std::get<static_cast<int>(propertyT::getKind())>(m_PropsList);
-    return PropHolder.getProp();
+  template <typename PropT>
+  typename std::enable_if<
+      std::is_base_of<detail::SimplePropertyBase, PropT>::value, bool>::type
+  has_property_helper() const {
+    const int PropKind = static_cast<int>(PropT::getKind());
+    if (PropKind >= detail::SimplePropKind::SimplePropKindSize)
+      return false;
+    return MSimpleProps[PropKind];
   }
 
-  template <typename propertyT> bool has_property() const {
-    if (static_cast<int>(propertyT::getKind()) >
-        property::detail::PropKind::PropKindSize)
-      return false;
-    return std::get<(int)(propertyT::getKind())>(m_PropsList).isInitialized();
+  template <typename PropT>
+  typename std::enable_if<
+      std::is_base_of<detail::ComplexPropertyBase, PropT>::value, bool>::type
+  has_property_helper() const {
+    const int PropKind = static_cast<int>(PropT::getKind());
+    for (const std::shared_ptr<detail::ComplexPropertyBase> &Prop :
```

Comment:
minor comment: this whole loop + 2 return statements could be replaced with "return std::any_of(...)"

Reply:
It would require including of <algorithm> which I wanted to avoid to make our headers more light-weighted.